lib/expressions/Term.ts

```diff
@@ -91,38 +96,57 @@ export class Literal<T> extends Term {
 }
 
 export class NumericLiteral extends Literal<number> {
-  private static readonly specificFormatters: {[key in C.PrimitiveNumericType]: (val: number) => string } = {
+  private readonly specificFormatter: ((val: number) => string);
+  public constructor(
+    public typedValue: number,
+    public typeURL: RDF.NamedNode,
+    public strValue?: string,
+    public language?: string,
+  ) {
+    super(typedValue, typeURL, strValue, language);
+    if (!typeCanBeProvidedTo(<LiteralTypes> typeURL.value, TypeAlias.SPARQL_NUMERIC)) {
+      throw new Error(
+        `TypeUrl '${this.type}' provided to NumericLiteral should implement ${TypeAlias.SPARQL_NUMERIC}`,
+      );
+    }
+
+    const type = this.type;
     // Avoid emitting non lexical integers
-    integer: value => value.toFixed(0),
-    float: value => value.toString(),
-    decimal: value => value.toString(),
-    // // Be consistent with float
-    // decimal: (value) => {
-    //   const jsDecimal = value.toString();
-    //   return jsDecimal.match(/\./)
-    //     ? jsDecimal
-    //     : jsDecimal + '.0';
-    // },
-
-    // https://www.w3.org/TR/xmlschema-2/#double
-    double(value) {
-      const jsExponential = value.toExponential();
-      const [ jsMantisse, jsExponent ] = jsExponential.split('e');
-
-      // Leading + must be removed for integer
-      // https://www.w3.org/TR/xmlschema-2/#integer
-      const exponent = jsExponent.replace(/\+/u, '');
-
-      // SPARQL test suite prefers trailing zero's
-      const mantisse = jsMantisse.includes('.') ?
-        jsMantisse :
-        `${jsMantisse}.0`;
-
-      return `${mantisse}E${exponent}`;
-    },
-  };
-
-  public type: C.PrimitiveNumericType;
+    if (typeCanBeProvidedTo(type, TypeURL.XSD_INTEGER)) {
+      this.specificFormatter = value => value.toFixed(0);
+    } else if (typeCanBeProvidedTo(type, TypeURL.XSD_DECIMAL)) {
+      this.specificFormatter = value => value.toString();
+    } else if (typeCanBeProvidedTo(type, TypeURL.XSD_FLOAT)) {
+      this.specificFormatter = value => value.toString();
+    } else if (typeCanBeProvidedTo(type, TypeURL.XSD_DOUBLE)) {
```

Comment:
The problem here is mostly runtime overhead I think. These are a lot of checks we need to for each number, which is a very common datatype.
If we have a subtypesFor, we can create a static map matching each concrete subtype to a concrete formatter a bit faster. Not sure if worth, just a gut feeling.

Reply:
Those are indeed a lot of checks for every number. would it help to make a function that would do these checks only when needed in `str()` ?

Reply:
It's better. But I still think a static map is the ideal approach here. Your choice ofc. You can also split the numeric class into the primitives (float, decimal, double), but that won't solve much.

You need either (derived type) -> (primitive type) -> (formatter), or (derived type) -> (formatter). There's plenty of options for building one I think.

Reply:
I don't think we can do a static map. Especially if we move sparqlee to have an open world mentality. Having a list of all subtypes would not be possible I think. Or just not scalable.

And like you said, splitting this class will not solve much. In that case we will need to do the check at construction like it is now. I'm sorry, I'll need to disappoint you. I will create a `specificFormattrCreator`.

Reply:
I don't think there is a scenario where the static maps don't scales but these checks do?

You are overestimating the likelihood of people providing _even more_ numeric subtypes, and sacrificing performance now for an unlikely scenario in a future world where sparqlee might be open world.

Even then, when it is open world, it would be preferable to have a static map, and a fallback mechanism for when the concrete type is not present.

Reply:
Maybe I just don't see it I'll make an issue about this and leave it with a creator?

Reply:
That's okay :)